frontend/src/lib/codemirror/useCompareExtension.ts

```diff
@@ -88,38 +45,55 @@ const diffGutter = gutter({
         }
     },
     lineMarkerChange(update) {
-        return update.docChanged || !compareNullableText(update.state.facet(targetText), update.startState.facet(targetText))
+        return update.docChanged || (update.state.facet(diffLineMap) != update.startState.facet(diffLineMap))
     },
     initialSpacer: () => marker,
 })
 
-export function useDelayedCompareExtension(viewRef: RefObject<EditorView>, compareTo: string, delayMs = 1000): Extension {
-    const editTime = useRef(0)
-    const timeSinceEdit = Date.now() - editTime.current
-    const timeout = useRef<any>()
-    const [, forceUpdate] = useState({})
-
-    return [
-        useCompareExtension(viewRef, timeSinceEdit > delayMs ? compareTo : undefined),
-        EditorState.transactionExtender.of(tr => {
-            if (tr.docChanged) {
-                editTime.current = Date.now()
-
-                if (timeout.current)
-                    clearTimeout(timeout.current)
-                timeout.current = setTimeout(() => {
-                    forceUpdate({})
-                }, delayMs)
-            }
-
-            return null
-        }),
-    ]
-}
+const createDiffWorker = createWorkerFactory(() => import("./useCompareExtension.worker"))
 
 // Extension that highlights lines in the doc that differ from `compareTo`.
 export default function useCompareExtension(viewRef: RefObject<EditorView>, compareTo: string): Extension {
     const [compartment] = useState(new Compartment())
+    const [diffLineMapCompartment] = useState(new Compartment())
+
+    const plugin = ViewPlugin.fromClass(class {
+        private worker = createDiffWorker()
+
+        constructor(private view: EditorView) {
+            this.updateDiff()
+        }
+
+        async update(update: ViewUpdate) {
+            if (update.docChanged) {
+                this.updateDiff()
+            }
+        }
+
+        async updateDiff() {
+            const diff = await this.worker.calculateDiff(this.view.state.facet(targetString), this.view.state.doc.toString())
+
+            // Convert diff changes to a map of line numbers -> change type
+            let map: DiffLineMap = {}
+
+            for (const [, , childStartLine, childEndLine] of diff) {
+                for (let i = childStartLine; i < childEndLine; i++) {
+                    map[i + 1] = marker
+                }
+            }
+
+            // Has our targetString been updated to a blank,
+            // and thus we should be showing no diff right now, while
+            // the view's been updating?
+            if (typeof this.view.state.facet(targetString) !== "string") {
+                map = {}
+            }
```

Comment:
When does this case occur? If the `targetString` is missing (null?) then wouldn't `diff` return `[]` and therefore `map` would be `{}`?

Reply:
tl;dr This is to handle the case where `compareTo` changes whilst `update` is executing.

During testing, it turned out that when the scratch is updated, the `update` function in useCompareExtension is fired before `isModified` is set to true (and thus `compareTo` is set to undefined). This means it would take two updates to get the diff to disappear instead of one.

I wasn't able to find a way to reorder the event handing to stop the diff calculation from occuring (and I'd love to remove this check if there is), but it turns out that await'ing the web worker reliably updates `compareTo` to its correct value - hence this check existing at the end of the function.